delocate/delocating.py

```diff
@@ -59,17 +71,46 @@ def delocate_tree_libs(lib_dict, lib_path, root_path):
         Filtered `lib_dict` dict containing only the (key, value) pairs from
         `lib_dict` where the keys are the libraries copied to `lib_path``.
     """
-    copied_libs = {}
-    delocated_libs = set()
+    # Test for errors first to avoid getting half-way through changing the tree
+    libraries_to_copy, libraries_to_delocate = _analyze_tree_libs(
+        lib_dict, root_path
+    )
+    # Copy libraries and update lib_dict.
+    lib_dict, copied_libraries = _copy_required_libs(
+        lib_dict, lib_path, root_path, libraries_to_copy
+    )
+    # Update the install names of local and copied libaries.
+    _update_install_names(
+        lib_dict, root_path, libraries_to_delocate | copied_libraries
+    )
+    return libraries_to_copy
+
+
+def _analyze_tree_libs(
+    lib_dict,  # type: Mapping[Text, Mapping[Text, Text]]
+    root_path  # type: Text
+):
+    # type: (...) -> Tuple[Dict[Text, Dict[Text, Text]], Set[Text]]
+    """ Verify then return which library files to copy and delocate.
+
+    Returns
+    -------
+    needs_copying : dict
+        The libraries outside of `root_path`.
+        This is in the `lib_dict` format since based on `delocate_tree_libs`
+        returns this value.
+    needs_delocating : set of str
+        The libraries inside of `root_path` which need to be delocated.
+    """
+    needs_delocating = set()  # Libraries which need install names updated.
+    needs_copying = {}  # A report of which libraries were copied.
     copied_basenames = set()
     rp_root_path = realpath(root_path)
-    rp_lib_path = realpath(lib_path)
-    # Test for errors first to avoid getting half-way through changing the tree
     for required, requirings in lib_dict.items():
-        if required.startswith('@'):  # assume @rpath etc are correct
-            # But warn, because likely they are not
-            warnings.warn('Not processing required path {0} because it '
-                          'begins with @'.format(required))
+        if required.startswith('@'):
+            # @rpath, etc, at this point is likely incorrect.
```

Comment:
`get_dependencies` can still return paths with `@` if they can't be resolved which is most likely to happen if a library dependency can't be found on any of the rpaths. If that happens then it's noted here and by `get_dependencies`.

So the question is should this be hard error rather than just logging it.  I think this should be treated like a missing library and raise `DelocationError`.

Reply:
Would you mind putting in a comment to say how the @ would reach this line?

For error vs logging - the question is - how to support the `delocate-listdeps` option, of showing all the library dependencies.  To be honest, I'm not sure what happens now if some of the depencies are missing, but I thought what happened is that the user gets warnings, but no error.  This is probably the right behavior for the `listdeps` option, but probably the wrong behavior for making a wheel - although it's possible someone somewhere is relying on missing dependencies passing silently.  We could make it a command line option, and a keyword argument to `tree_libs`, and `wheel_libs`.   Or make `_analyze_tree_libs` public and use that instead, with a keyword argument.  What do you think?

Reply:
> although it's possible someone somewhere is relying on missing dependencies passing silently. We could make it a command line option

I agree with making this a command line option just in case.

> For error vs logging - the question is - how to support the delocate-listdeps option, of showing all the library dependencies.

While it's related, I think `delocate-listdeps` is outside the scope of this PR, which was already so big that I had to cull a lot of changes to make it presentable.

I think the current tests assume the `delocate-listdeps` does not look up libraries recursively.  I had to leave `delocate-listdeps` using the old `tree_libs` code path for at the moment, but it should be possible to switch it to use `walk_directory` later.

Reply:
Requirements starting with `@` can no longer reach that section of the code.  They are always filtered out or resolved by other functions before going into `lib_dict`.

`get_dependencies` does not return invalid paths anymore, and it's up to callers of `get_dependencies` to handle what to do when libraries are missing.

Dependencies can now be ignored from the command line.  This is handled by the `delocate_path` function.